stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+    return 0;
+  }
+
+  /**
+   * Implements the function of type CVDlsJacFn which is the
+   * user-defined callback for CVODES to calculate the jacobian of the
+   * ode_rhs wrt to the states y. The jacobian is stored in column
+   * major format.
+   */
+  constexpr static int cv_jacobian_states(realtype t, N_Vector y, N_Vector fy,
+                                          SUNMatrix J, void* user_data,
+                                          N_Vector tmp1, N_Vector tmp2,
+                                          N_Vector tmp3) {
+    const cvodes_integrator_adjoint_vari* integrator
+        = static_cast<const cvodes_integrator_adjoint_vari*>(user_data);
+    integrator->jacobian_states(t, y, J);
+    return 0;
+  }
+
+  /**
+   * Implements the CVLsJacFnB function for evaluating the jacobian of
+   * the adjoint problem.
+   */
+  constexpr static int cv_jacobian_adj(realtype t, N_Vector y, N_Vector yB,
+                                       N_Vector fyB, SUNMatrix J,
+                                       void* user_data, N_Vector tmp1,
+                                       N_Vector tmp2, N_Vector tmp3) {
+    const cvodes_integrator_adjoint_vari* integrator
+        = static_cast<const cvodes_integrator_adjoint_vari*>(user_data);
+    integrator->jacobian_adj(t, y, J);
+    return 0;
+  }
+
+  /**
+   * Calculates the ODE RHS, dy_dt, using the user-supplied functor at
+   * the given time t and state y.
+   */
+  inline void rhs(double t, const double y[], double dy_dt[]) const {
+    const Eigen::VectorXd y_vec = Eigen::Map<const Eigen::VectorXd>(y, N_);
+
+    const Eigen::VectorXd dy_dt_vec = rhs(t, y_vec, value_of_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", dy_dt_vec.size(),
+                     "states", N_);
+
+    Eigen::Map<Eigen::VectorXd>(dy_dt, N_) = dy_dt_vec;
+  }
+
+  /*
+   * Calculate the adjoint sensitivity RHS for varying initial conditions
+   * and parameters
+   *
+   * Equation 2.23 in the cvs_guide.
+   *
+   * @param[in] initial var vector
+   * @param[in] param var vector
+   * @param[in] t time
+   * @param[in] y state of the base ODE system
+   * @param[in] yB state of the adjoint ODE system
+   * @param[out] yBdot evaluation of adjoint ODE RHS
+   */
+  inline void rhs_adj_sens(double t, N_Vector y, N_Vector yB,
+                           N_Vector yBdot) const {
+    Eigen::Map<const Eigen::VectorXd> y_vec(NV_DATA_S(y), N_);
+    Eigen::Map<Eigen::VectorXd> mu(NV_DATA_S(yB), N_);
+    Eigen::Map<Eigen::VectorXd> mu_dot(NV_DATA_S(yBdot), N_);
+    mu_dot.setZero();
+
+    const nested_rev_autodiff nested;
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(y_vec);
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = rhs(t, y_vars, value_of_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", f_y_t_vars.size(),
+                     "states", N_);
+
+    f_y_t_vars.adj() = -mu;
+
+    grad();
+
+    mu_dot = y_vars.adj();
+  }
```

Comment:
```suggestion
  inline void rhs_adj_sens(double t, N_Vector y, N_Vector yB,
                           N_Vector yBdot) const {

    const nested_rev_autodiff nested;
    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(Eigen::Map<const Eigen::VectorXd>(NV_DATA_S(y), N_));
    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars = this->rhs(t, y_vars, value_of_args_tuple_);
    check_size_match(solver_->function_name_, "dy_dt", f_y_t_vars.size(),
                     "states", N_);
    f_y_t_vars.adj() = -Eigen::Map<Eigen::VectorXd>(NV_DATA_S(yB), N_);
    grad();
    Eigen::Map<Eigen::VectorXd>(NV_DATA_S(yBdot), N_) = y_vars.adj();
  }
```

Reply:
Using `this->rhs` feels odd to me, really. I would prefer to not prefix it with this. I mean, this is standard to call functions of the own instance. One would have to prefix everything with `this` in this logic, which does not seem to be necessary.